src/ohio/ext/numpy.py

```diff
@@ -0,0 +1,212 @@
+    # NOTE: intermediary between the source (array) and the target (database).
+    # NOTE: The fmt is relatively inconsequential so long as the database
+    # NOTE: schema enforces the proper typing and precision.
+    #
+    # NOTE: Rather, fmt matters most to size of the generated CSV; but, at that
+    # NOTE: level, it's just an optimization parameter. (We'd want to keep
+    # NOTE: floats long by default, to avoid losing precision, but allow user
+    # NOTE: to shorten as they like via custom fmt, for smaller CSV payloads.)
+    with ohio.pipe_text(numpy.savetxt,
+                        arr,
+                        delimiter=',',
+                        **savetxt_kwargs) as pipe, \
+            connectable.begin() as tx:
+        # connectable is either an Engine or a Connection,
+        # and as such tx is either a new Connection or the Transaction
+        conn = tx if hasattr(tx, 'execute') else connectable
+        cursor = conn.connection.cursor()
+        cursor.copy_expert(sql, pipe)
+
+
+def pg_copy_from_query(query, connectable, dtype):
+    """Construct ``array`` from database ``query`` via PostgreSQL
+    ``COPY``.
+
+    ``ohio.PipeTextIO`` enables the in-process "piping" of the
+    PostgreSQL ``COPY`` command into NumPy's ``fromiter``, for quick,
+    memory-efficient construction of ``array`` from database, (and
+    without the needless involvement of the local file system).
+
+    For example, given a SQLAlchemy ``connectable`` – either a database
+    connection ``Engine`` or ``Connection``::
+
+        >>> from sqlalchemy import create_engine
+        >>> engine = create_engine('postgresql://')
+
+    We may construct a NumPy ``array`` from a given query::
+
+        >>> arr = pg_copy_from_query(
+        ...     'select value0, value1, value3 from data',
+        ...     engine,
+        ...     float,
+        ... )
+
+    """
+    return _pg_copy_from("({})".format(query), connectable, dtype)
+
+
+def pg_copy_from_table(table_name, connectable, dtype, columns=None):
+    """Construct ``array`` from database table via PostgreSQL ``COPY``.
+
+    ``ohio.PipeTextIO`` enables the in-process "piping" of the
+    PostgreSQL ``COPY`` command into NumPy's ``fromiter``, for quick,
+    memory-efficient construction of ``array`` from database, (and
+    without the needless involvement of the local file system).
+
+    For example, given a SQLAlchemy ``connectable`` – either a database
+    connection ``Engine`` or ``Connection``::
+
+        >>> from sqlalchemy import create_engine
+        >>> engine = create_engine('postgresql://')
+
+    We may construct a NumPy ``array`` from the contents of a specified
+    table::
+
+        >>> arr = pg_copy_from_table(
+        ...     'data',
+        ...     engine,
+        ...     float,
+        ... )
+
+    """
+    source = _sql_table_columns(table_name, columns)
+    return _pg_copy_from(source, connectable, dtype)
+
+
+# Internals #
+
+class NpCsvReader:
```

Comment:
I'm not sure shortening 'Numpy' to 'Np' is really needed or helpful here, though it's not too consequential as it's an internal class

Reply:
@thcrock Thanks!

Yeah, I'd say the main thing is it's just an internal class.

FYI, that is _some_ sort of standard for NumPy, for some reason 🤷‍♂️ 

[(numpy docs)](https://docs.scipy.org/doc/numpy/user/basics.types.html)

![image](https://user-images.githubusercontent.com/530998/56751733-5f90a080-674c-11e9-9fc1-3ae49e0d16d2.png)